lib/ood_core/batch_connect/templates/vnccontainer.rb

```diff
@@ -0,0 +1,248 @@
+                kill -0 ${VNC_PID} 2>/dev/null && break
+              done
+
+              # If we fail to start it after so many tries, then just give up
+              kill -0 ${VNC_PID} 2>/dev/null || clean_up 1
+
+              # Parse output for ports used
+              display=$(echo "${VNC_OUT}" | awk -F':' '/^Desktop/{print $NF}')
+              port=$((5900+display))
+
+              echo "Successfully started VNC server on ${host}:${port}..."
+
+              #{super}
+            EOT
+          end
+
+          # Run the script under the VNC server's display
+          def run_script
+            %(DISPLAY=:${display} #{super})
+          end
+
+          # After startup the main script, scan the VNC server log file for
+          # successful connections so that the password can be reset
+          def after_script
+            websockify_cmd = context.fetch(:websockify_cmd, "${WEBSOCKIFY_CMD:-/opt/websockify/run}").to_s
+
+            <<-EOT.gsub(/^ {14}/, "")
+              #{super}
+
+              # Launch websockify websocket server
+              module load #{container_module}
+              echo "Starting websocket server..."
+              websocket=$(find_port)
+              #{container_command} exec instance://#{instance_name} #{websockify_cmd} -D ${websocket} localhost:${port}
+
+              # Set up background process that scans the log file for successful
+              # connections by users, and change the password after every
+              # connection
+              echo "Scanning VNC log file for user authentications..."
+              while read -r line; do
+                if [[ ${line} =~ "Full-control authentication enabled for" ]]; then
+                  change_passwd
+                  create_yml
+                fi
+              done < <(tail -f --pid=${SCRIPT_PID} "#{vnc_log}") &
+            EOT
+          end
+
+          # Clean up the running VNC server and any other stale VNC servers
+          def clean_script
+            <<-EOT.gsub(/^ {14}/, "")
+              #{super}
+              module load #{container_module}
+
+              #{vnc_clean}
+              [[ -n ${display} ]] && vncserver -kill :${display}
+              #{container_command} instance stop #{instance_name}
+            EOT
+          end
+
+          # Log file for VNC server
+          def vnc_log
+            context.fetch(:vnc_log, "vnc.log").to_s
+          end
+
+          # Password file for VNC server
+          def vnc_passwd
+            context.fetch(:vnc_passwd, "vnc.passwd").to_s
+          end
+
+          def instance_name 
+            context.fetch(:instance_name, "desktop").to_s
+          end
+
+          def container_module 
+            context.fetch(:container_module, "singularity").to_s
+          end
+
+          def container_command 
+            context.fetch(:container_command, "singularity").to_s
+          end
+
+          # Arguments sent to `vncserver` command
+          def vnc_args
```

Comment:
Seems like we have to override a lot of things, and that's fine, but I wonder if we inherited from `VNC` instead of the parent `Template` we wouldn't need to duplicate this?

Reply:
That's fine, I can refactor to do that.

Reply:
I ultimately decided not to refactor this because `super` for `VNC` pulled in all sorts of baggage and was more trouble to sort out than it was worth.